package/MDAnalysis/topology/TOPParser.py

```diff
@@ -308,3 +519,59 @@ def parsesection_mapper(self, atoms_per, numlines, mapper):
                 if val:
                     section.append(mapper(val))
         return section
+
+    def parse_dihedrals(self, diha, dihh):
+        """Combines hydrogen and non-hydrogen containing AMBER dihedral lists
+        and extracts sublists for conventional dihedrals and improper angles
+
+        Parameters
+        ----------
+        diha : list of 4 np.int64 element tuples
+            The dihedrals not involving hydrogens
+        dihh : list of 4 np.int64 element tuples
+            The dihedrals involving hydrogens
+
+        Returns
+        -------
+        dihedrals : :class:`Dihedrals`
+            A :class:`Dihedrals` instance containing a list of all unique
+            dihedrals as defined by the parm7 file
+        impropers : :class:`Impropers`
+            A :class:`Impropers` instance containing a list of all unique
+            improper dihedrals as defined by the parm7 file
+
+        Note
+        ----
+        As detailed in http://ambermd.org/formats.html, the dihedral sections
+        of parm7 files contain information about both conventional dihedrals
+        and impropers. The following must be accounted for:
+        1) If the fourth atom in a dihedral entry is given a negative value,
+        this indicates that it is an improper.
+        2) If the third atom in a dihedral entry is given a negative value,
+        this indicates that it 1-4 NB interactions are ignored for this
+        dihedrals. This could be due to the dihedral within a ring, or if it is
+        part of a multi-term dihedral definition. To account for the latter
+        case and avoid the possibility of duplicate dihedral entries, we check
+        that any such dihedral does not already exist in the accumulated
+        dihderals list.
+        Caveat: currently we assume that multiterm dihedrals are
+        added to the parm7 file after the initial dihedral. This is likely true
+        for leap generated files, but such behaviour is not necessarily
+        guaranteed.
+        """
+        improp = []
+        dihed = []
+        for i in itertools.chain(diha, dihh):
+            if i[3] < 0:
+                improp.append(i[:3]+(abs(i[3]),))
+            elif i[2] < 0:
+                vals = i[:2] + (abs(i[2]),) + i[3:]
+                if vals in dihed:
```

Comment:
nitpicky, but here you could just have `if not vals in dihed:` and not have the `continue` branch

Reply:
That does make a lot cleaner thanks :)

Unfortunately, on profiling my original proposed changes to the TOPParser, and it seems that there is a very large overhead associated with this particular if statement (switching to `if not vals in dihed` unfortunately does not improve the issue). When loading a ~ 1 million atom system (AMBER benchmark's STMV system), I see a 50x decrease in performance, taking ~ 1850 seconds to create a universe vs ~ 36 seconds without the if statement.

The original purpose of this particular line was to prevent duplicate dihedral entries which would arise from multiterm dihedrals. However, I am no longer sure if it is actually necessary. Looking more closely at the PSF reader, it seems to not attempt to remove extra multiterm dihedrals either. Additionally, it appears that duplicate dihedrals are removed somewhere in `topologyobjects.py` (although I am not 100% exactly where it is happening). 

I was therefore wondering if it would be reasonable to allow for duplicate dihedrals on read, or if this might cause some odd behaviour somewhere?

Reply:
hmm, you could make dihed and improp a set, it's hard to beat the performance of Python's inbuilt set.

Can you link the STMV system you mentioned?

Reply:
@richardjgowers 
Thanks, as per your suggestion I ended up doing a sorted set conversion in 308bb68 which ended up with much lower overheads ( ~ 1 s extra compared to no duplicate removal for the creation of the STMV system universe).

I did also get rid of the np.int64 conversion for the bond reader (under parse_bonded), as it had a rather high cost and (hopefully I've not missed anything) as far as I can tell the topology attribute for bonds doesn't require np.ints.

Unfortunately the code in this PR does still double the cost of the TOPParser (STMV universe creation goes from ~ 14 s to ~ 28 s) but I think that is unavoidable.

With regards to the STMV system, it can be found as part of the AMBER benchmark suite (under PME/STMV_production_NPT_4fs): http://ambermd.org/Amber16_Benchmark_Suite.tar.bz2